### PR TITLE
fix(cpo, orpo): truncate responses independently to prevent empty completions

### LINKED ISSUE #6548: [CPO/SimPO] Truncation in CPOTrainer can silently produce empty completions -> NaN loss with loss_type="simpo"
## Description

In `trl/experimental/cpo/cpo_trainer.py`, the response truncation logic in
`tokenize_row` truncates **both** responses using the length of the **longer**
response of the pair:

```python
longer_response_length = max(len(chosen_tokens["input_ids"]), len(rejected_tokens["input_ids"]))

# if combined sequence is too long, truncate the response
for answer_tokens in [chosen_tokens, rejected_tokens]:
    if len(answer_tokens["prompt_input_ids"]) + longer_response_length > self.max_length:
        for k in ["input_ids", "attention_mask"]:
            answer_tokens[k] = answer_tokens[k][: self.max_length - longer_response_length]
```

(https://github.com/huggingface/trl/blob/4708879046dde897bdb84667a97e781bfc1965d2/trl/experimental/cpo/cpo_trainer.py#L554-L560)

When `longer_response_length > self.max_length`, the slice bound
`self.max_length - longer_response_length` is **negative**, so `lst[:negative]`
cuts from the end. If the *shorter* response of the pair has fewer tokens than
that negative offset, it is truncated to an **empty sequence** (0 completion
tokens, all labels are -100).

Consequences:

- With `loss_type="simpo"`, log-probs are length-normalized, so an empty
  completion produces a `0/0` division -> `NaN` in `logps`, which propagates to
  the loss and destroys all model weights after the first optimizer step.
  Training metrics show `loss = 0`, `grad_norm = nan`, all rewards/logps `nan`.
- With other loss types the empty completion contributes a silent `logp = 0`,
  which is incorrect but does not crash — arguably worse, since it corrupts
  training without any visible symptom.

This is easy to hit in practice with self-generated preference data
(e.g., math reasoning), where response lengths are heavy-tailed: a single pair
with one very long response (> max_length) and one normal-length response
triggers it.

## Reproduction

Pure-Python demonstration of the slicing logic (3 lines):

```python
max_length = 2048
chosen = list(range(5000))    # long response, > max_length
rejected = list(range(600))   # normal short response

longer = max(len(chosen), len(rejected))          # 5000
print(len(chosen[: max_length - longer]))         # 2048  (truncated, survives)
print(len(rejected[: max_length - longer]))       # 0     (silently EMPTY)
```

End-to-end reproduction through the trainer (any small public model):

```python
import torch
from datasets import Dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl.experimental.cpo import CPOTrainer, CPOConfig

model_id = "Qwen/Qwen2.5-0.5B"
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float32)

pair = {
    "prompt": "Question: what is 2+2?\nAnswer:",
    "chosen": " step" * 5000,          # ~5000 tokens, exceeds max_length
    "rejected": " The answer is 4.",   # ~7 tokens, healthy short response
}

def count_completion_tokens(max_length):
    args = CPOConfig(
        output_dir="/tmp/cpo_repro",
        loss_type="simpo",
        cpo_alpha=0.0,
        beta=0.1,
        max_length=max_length,
        per_device_train_batch_size=1,
        report_to="none",
    )
    trainer = CPOTrainer(
        model=model,
        args=args,
        train_dataset=Dataset.from_list([pair]),
        processing_class=tokenizer,
    )
    row = trainer.train_dataset[0]
    n_chosen = sum(t != -100 for t in row["chosen_labels"])
    n_rejected = sum(t != -100 for t in row["rejected_labels"])
    return n_chosen, n_rejected

print("max_length=2048:", count_completion_tokens(2048))
# -> (2048, 0)   rejected is EMPTY -> NaN loss with loss_type="simpo"

print("max_length=8192:", count_completion_tokens(8192))
# -> (5001, 7)   control: everything fits, both responses intact
```

Same data, only `max_length` differs: the short response is destroyed exactly
when the *other* response exceeds `max_length`.

## Expected behavior

Each response should be truncated based on **its own** length, e.g.:

```python
for answer_tokens in [chosen_tokens, rejected_tokens]:
    if len(answer_tokens["prompt_input_ids"]) + len(answer_tokens["input_ids"]) > self.max_length:
        for k in ["input_ids", "attention_mask"]:
            answer_tokens[k] = answer_tokens[k][
                : max(0, self.max_length - len(answer_tokens["prompt_input_ids"]))
            ]
```

so that a short response is never emptied because its pair partner is long.
Additionally, a guard (warning or error) when a completion ends up with zero
tokens after truncation would make failures like the SimPO NaN immediately
diagnosable instead of silent.

### System Info

- Platform: Linux-6.8.0-124-generic-x86_64-with-glibc2.39
- Python version: 3.12.3
- TRL version: 1.6.0
- PyTorch version: 2.11.0+cu130
- accelerator(s): NVIDIA L40S, NVIDIA L40S
- Transformers version: 5.12.1
- Accelerate version: 1.14.0
- Accelerate config: 
  - compute_environment: LOCAL_MACHINE
  - distributed_type: FSDP
  - mixed_precision: bf16
  - use_cpu: False
  - debug: False
  - num_processes: 2
  - machine_rank: 0
  - num_machines: 1
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - fsdp_config: {'fsdp_activation_checkpointing': False, 'fsdp_auto_wrap_policy': 'TRANSFORMER_BASED_WRAP', 'fsdp_backward_prefetch': 'BACKWARD_PRE', 'fsdp_cpu_ram_efficient_loading': True, 'fsdp_forward_prefetch': False, 'fsdp_offload_params': False, 'fsdp_sharding_strategy': 'FULL_SHARD', 'fsdp_state_dict_type': 'SHARDED_STATE_DICT', 'fsdp_sync_module_states': True, 'fsdp_use_orig_params': False, 'fsdp_version': 1}
  - downcast_bf16: yes
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 5.0.0
- HF Hub version: 1.23.0
- bitsandbytes version: 0.49.2
- DeepSpeed version: not installed
- Liger-Kernel version: not installed
- PEFT version: 0.19.1
- vLLM version: 0.23.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by RohanMali2003**
# What does this PR do?

Fixes #6548 

This PR implements the first half of the fix for the CPO/SimPO truncation bug outlined in #6548. It updates both `CPOTrainer` and `ORPOTrainer` to truncate responses using their own independent budget rather than the length of the longer response. 

**Changes:**
- Replaced `longer_response_length` with independent `len(answer_tokens["prompt_input_ids"])` calculation.
- Added a `max(0, ...)` bound to prevent Python's negative slicing from silently clearing shorter responses.
- Added a `logger.warning` to alert the user if a completion is entirely truncated, which previously caused silent NaN losses with SimPO.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

(i used an AI coding tool to review my code and ensure that the core logic matched the issue's requirements and was safely applied to both trainers- cpo and orpo, and to help with documentation. Suggestions from @ErenAta16 greatly helped, and he correctly pointed out the issue being duplicated in orpo.)

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes tokenization/truncation in core CPO/ORPO training paths, which can alter sequence lengths and gradients; scope is localized and covered by a new regression test.
> 
> **Overview**
> Fixes preference-pair truncation in **CPOTrainer** and **ORPOTrainer** so chosen and rejected completions are capped using each branch’s own prompt+response budget instead of sharing the longer response’s length.
> 
> Truncation now uses `max(0, max_length - len(prompt))` per answer, avoiding negative slices that could wipe the shorter completion and trigger **SimPO** NaNs. A one-time **warning** is logged when truncation leaves an empty completion.
> 
> Adds **`test_simpo_loss_is_finite_with_truncated_completion`**, which trains one SimPO step with a very long chosen completion under a tight `max_length` and asserts finite loss, grad norm, and model parameters.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 49d060cb062edc182e092502fd7d03861c174533. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by YAO-001** [comment]
Thanks @RohanMali2003 — I accepted the invitation and pushed commit `278cf743` directly to `fix-cpo-truncation`.

This adds:

- post-tokenization filtering for decoder-only CPO and ORPO train/eval datasets when either side has no supervised completion tokens;
- matching parameterized regression tests for independent truncation in both trainers;
- train/eval coverage for fully truncated pair removal;
- a one-step CPO SimPO regression test checking finite loss, gradient norm, and model parameters.

Validation:

- before adding the filtering logic, the two new filtering tests failed as expected while the other 9 targeted cases passed;
- after the change, all 11 targeted cases passed;
- the complete CPO/ORPO test files passed: `41 passed`;
- `ruff check`, `ruff format`, and the doc-builder style check all passed.

I kept your core truncation and warning changes intact. Please take a look when you have time; I'm happy to adjust anything needed before the draft is marked ready for review.

**Comment by RohanMali2003** [comment]
Hi @YAO-001, this looks absolutely fantastic! 

I just reviewed the commits and the implementation is spot on. Thank you for running the formatting and linting checks locally, too! I've just marked the PR to be ready for review. It was fun collaborating with you! :)

**Review by RohanMali2003 — COMMENTED** [review]
All good🤞

**Comment by RohanMali2003** [comment]
Hi @albertvillanova @qgallouedec ! I saw you guys are active on this repository, so I wanted to ping you- It would be cool if either of you could review this PR.

The truncation bug has been fixed and we've added the corresponding dataset filtering (mirroring `DPOTrainer`) along with the parameterized regression tests. 

Could one of you please approve the workflow runs so the CI can execute when you get a chance? Thanks a ton!

**Comment by qgallouedec** [comment]
Thanks for the detailed issue + fix.
Confirming this is a real regression: #4965/#4966 removed `max_prompt_length` and substituted `longer_response_length` into the slice bound, wrong quantity, a response's budget is `max_length - len(prompt)`. It over-truncates on every trigger, not just the NaN case.

The truncation fix in both trainers is what I want. Could you trim the rest?

- Drop the dataset filtering. Once the bound is fixed, an empty completion only happens when the prompt alone fills `max_length`
- Then `warning_once` and the filter are redundant; keep one.
- Tests: `Trainer.__new__` + the `_CharacterTokenizer` stub isn't a pattern we use; a small parametrized case per trainer is enough. Even no test is fine at this point

**Comment by RohanMali2003** [comment]
Okay, I'll do that. Thanks for the review! :))

**Comment by RohanMali2003** [comment]
Hey, @qgallouedec , I've just pushed a commit that removes the `.filter()` logic and the extra tokenizer/parameterized tests as you asked in the prev review. I trimmed the PR down to just the core independent truncation fix and `logger.warning_once()` for both `CPOTrainer` and `ORPOTrainer`. 

Let me know if there's anything else you need changed!

**Comment by qgallouedec** [comment]
Thanks @RohanMali2003, the truncation fix looks solid and I confirmed the new SimPO regression test passes.

One loose end before this is mergeable: `test_orpo_trainer.py` still imports `Dataset` from the earlier version of the PR but no longer uses it, which will fail the code quality check.

https://github.com/huggingface/trl/blob/3b97679aecb18bb660e76e51cd83e181af8e3506/tests/experimental/test_orpo_trainer.py#L17

Could you drop that import?

**Comment by RohanMali2003** [comment]
Absolutely! I'll do it ASAP

**Comment by RohanMali2003** [comment]
@qgallouedec All done! I've removed the unused Dataset import as requested. Let me know if everything looks good to go!

**Comment by RohanMali2003** [comment]
@qgallouedec All done! I've removed the unused Dataset import from `test_orpo_trainer` as requested. Let me know if everything looks good to go!`

**Comment by RohanMali2003** [comment]
@qgallouedec Removed the unused Dataset import from `test_orpo_trainer.py`.
